File: daily/287easy/kaprekar.py
#!/usr/bin/env python3

def largest_digit(n):
    return int(max(str(n)))


def desc_digits(n):
    return int(''.join(sorted('{:04}'.format(n), reverse=True)))


def asc_digits(n):
    return int(''.join(sorted('{:04}'.format(n))))


def kaprekar(n):
    i = 0

    while n != 6174:
        n = desc_digits(n) - asc_digits(n)
        i += 1

    return i


if __name__ == '__main__':
    print('largest_digit(1234) -> {}'.format(largest_digit(1234)))
    print('largest_digit(3253) -> {}'.format(largest_digit(3253)))
    print('largest_digit(9800) -> {}'.format(largest_digit(9800)))
    print('largest_digit(3333) -> {}'.format(largest_digit(3333)))
    print('largest_digit(120) -> {}\n'.format(largest_digit(120)))

    print('desc_digits(1234) -> {}'.format(desc_digits(1234)))
    print('desc_digits(3253) -> {}'.format(desc_digits(3253)))
    print('desc_digits(9800) -> {}'.format(desc_digits(9800)))
    print('desc_digits(3333) -> {}'.format(desc_digits(3333)))
    print('desc_digits(120) -> {}\n'.format(desc_digits(120)))

    print('kaprekar(6589) -> {}'.format(kaprekar(6589)))
    print('kaprekar(5455) -> {}'.format(kaprekar(5455)))
    print('kaprekar(6174) -> {}'.format(kaprekar(6174)))

File: daily/287easy/kaprekar.c
#include <stdio.h>

char a[4];

char largest_digit(short n) 
{
	char m = n / 1000;

	if (m < n % 1000 / 100)
		m = n % 1000 / 100;

	if (m < n % 100 / 10)
		m = n % 100 / 10;

	return (m < n % 10) ? n % 10 : m;
}

void sort_digits(short n)
{
	char i, j;

	a[0] = n / 1000;
	a[1] = n % 1000 / 100;
	a[2] = n % 100 / 10;
	a[3] = n % 10;

	for (i = 0; i < 3; i++)
		for (j = i + 1; j < 4; j++)
			if (a[i] < a[j]) {
				a[i] += a[j];
				a[j] = a[i] - a[j];
				a[i] -= a[j];
			}
}

short desc_digits(short n)
{
	sort_digits(n);

	return a[0] * 1000 + a[1] * 100 + a[2] * 10 + a[3];
}

short asc_digits(short n)
{
	sort_digits(n);

	return a[3] * 1000 + a[2] * 100 + a[1] * 10 + a[0];
}

unsigned long long kaprekar(short n)
{
	char i;

	for (i = 0; n ^ 6174; i++)
		n = desc_digits(n) - asc_digits(n);

	return i;
}

int main()
{
	printf("largest_digit(1234) -> %hhd\n", largest_digit(1234));
	printf("largest_digit(3253) -> %hhd\n", largest_digit(3253));
	printf("largest_digit(9800) -> %hhd\n", largest_digit(9800));
	printf("largest_digit(3333) -> %hhd\n", largest_digit(3333));
	printf("largest_digit(120) -> %hhd\n\n", largest_digit(120));

	printf("desc_digits(1234) -> %hd\n", desc_digits(1234));
	printf("desc_digits(3253) -> %hd\n", desc_digits(3253));
	printf("desc_digits(9800) -> %hd\n", desc_digits(9800));
	printf("desc_digits(3333) -> %hd\n", desc_digits(3333));
	printf("desc_digits(120) -> %hd\n\n", desc_digits(120));

	printf("kaprekar(6589) -> %lld\n", kaprekar(6589));
	printf("kaprekar(5455) -> %lld\n", kaprekar(5455));
	printf("kaprekar(6174) -> %lld\n", kaprekar(6174));

	return 0;
}
